docker: fix docker image to suit storage in memory

ZhouXing19 opened the pull request:
Currently the datasets created in the init process are stored on disk,
and they are inaccessible if the “real” server is started with storage in memory.

This is because, for the container initialization, the current logic is:
1. start_init_server: start the single-node server in the background, with
default on-disk storage setting;
2. run_init_queries: create default db and user, and run the init scripts
saved in
docker-entrypoint-initdb.d;
3. *stop the init server and restart the server* with given parameters in foreground.

In this commit we start and keep running the init server with the given
parameters, and eliminate the “stop and restart” step. Also, bring the server
process from background to foreground with `fg` once all init queries are finished.

Fixes #80005.

Release note (docker): refactor the initialization process of the docker image
to accomodate initialization scripts with memory storage.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80036)
<!-- Reviewable:end -->

otan commented on pkg/testutils/docker/single_node_docker_test.go:
```diff
@@ -499,17 +523,17 @@ func (dn *dockerNode) waitServerStarts(ctx context.Context) error {
 		return nil
 	}
-	return errors.Wrap(errors.Newf("%s", res), "error in waiting the server to start")
+	return errors.Wrap(errors.Newf("%#v", res), "error in waiting the initialization to finish")
```
nit: `errors.New("error waiting for initialization to finish: %v", err)`

otan reviewed (approved):
Lookgs good. You'll need to fix up the release note + remove [WIP] from the commit message.

We'll also want to backport this.

ZhouXing19 committed:
docker: fix docker image to suit the memory storage

Currently the datasets created in the init process are stored on disk,
and they are inaccessible if the “real” server is started with storage in memory.

This is because for the container initialization, the current logic is:
1. start_init_server: start the single-node server in the background, with
default on-disk storage setting;
2. run_init_queries: create default db and user, and run the init scripts
saved in
docker-entrypoint-initdb.d;
3. *stop the init server and restart the server* with given parameters in foreground.

In this commit we start and keep running the init server with the given
parameters, and eliminate the “stop and restart” step. Also, bring the server
process from background to foreground with fg once all init queries are finished.

Fixes #80005.

Release note (docker): refactor the initialization process of the docker image
to accomodate the use case with memory storage.

rafiss commented on build/deploy/cockroach.sh:
```diff
@@ -257,64 +251,28 @@ create_default_user() {
+    echo "init_finished" > ./cockroach-data/init_success
+    fg %1
+  else
+      # Start the real server.
```
"real server" is probably not a necessary comment -- since the `if` branch above is also a "real" server now

ZhouXing19 replied:
Done.

rafiss reviewed (approved):
lgtm!

ZhouXing19 committed:
list files

ZhouXing19 commented:
Thanks!

bors r+
